Updated sidecar to not require VolumeSnapshotClass for snapshot deletion

huffmanca opened the pull request:
**What type of PR is this?**
/kind flake

**What this PR does / why we need it**:
Allows the sidecar to delete VolumeSnapshots if the VolumeSnapshotClass has already been deleted. This PR is a followup to https://github.com/kubernetes-csi/external-snapshotter/pull/275 for the sidecar.

**Which issue(s) this PR fixes**:
Fixes #264 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Allows the sidecar to delete volume snapshots if the volume snapshot class is not found.
```

huffmanca commented:
@xing-yang ,

I'm submitting this as a WIP, as I can't get it to fully past the delete tests. There's an issue with the expected number of delete events on these tests, where we're expecting 0 events, and the array of events is nil. I haven't been able to trace why this is the cause. Would you have any thoughts?

xing-yang commented:
/ok-to-test

xing-yang commented:
@huffmanca I think it is because this PR allows snapshot to be deleted when snapshot class is not found.  Check if those failed tests have a snapshot class that is not found.

humblec commented:
@xing-yang is it valid to delete volumesnapshot if subjected SC not found? at times, we may require some params or data from SC to perform delete operation, Isnt it ?

xing-yang commented:
> @xing-yang is it valid to delete volumesnapshot if subjected SC not found? at times, we may require some params or data from SC to perform delete operation, Isnt it ?

Yes, it is valid to delete snapshot even when SC is already deleted.  We only need to have secrets when deleting the snapshot which is already added as annotation to the content.

xing-yang commented:
@huffmanca can you rebase your PR and see if you still have the issue with unit tests?

xing-yang commented on pkg/sidecar-controller/snapshot_delete_test.go:
```diff
@@ -334,16 +334,16 @@ func TestDeleteSync(t *testing.T) {
-			expectedDeleteCalls: []deleteCall{{"sid1-15", nil, nil}},
-			test:                testSyncContent,
-		},
+		}, /*
```
Uncomment out this test and make it a positive test.

xing-yang commented on pkg/sidecar-controller/snapshot_delete_test.go:
```diff
@@ -334,16 +334,16 @@ func TestDeleteSync(t *testing.T) {
-		},
+		}, /*
+			{
+				name:                "1-15 - (dynamic)deletion of content with no snapshotclass should produce error",
```
s/should produce error/should succeed

xing-yang commented on pkg/sidecar-controller/snapshot_delete_test.go:
```diff
@@ -334,16 +334,16 @@ func TestDeleteSync(t *testing.T) {
+				name:                "1-15 - (dynamic)deletion of content with no snapshotclass should produce error",
+				initialContents:     newContentArrayWithDeletionTimestamp("content1-15", "sid1-15", "snap1-15", "sid1-15", "", "", "snap1-15-volumehandle", deletePolicy, nil, &defaultSize, true, &timeNowMetav1),
+				expectedContents:    newContentArrayWithDeletionTimestamp("content1-15", "sid1-15", "snap1-15", "sid1-15", "", "", "snap1-15-volumehandle", deletePolicy, nil, &defaultSize, true, &timeNowMetav1),
+				expectedEvents:      []string{"Warning SnapshotDeleteError"},
```
Remove this line.

xing-yang commented on pkg/sidecar-controller/snapshot_delete_test.go:
```diff
@@ -334,16 +334,16 @@ func TestDeleteSync(t *testing.T) {
+				errors:              noerrors,
+				expectedDeleteCalls: []deleteCall{{"sid1-15", nil, nil}},
+				test:                testSyncContent,
+			},*/
```
For test 1-4, add the following line after "expectedEvents:   noevents,":
    expectedDeleteCalls: []deleteCall{{"sid1-1", nil, fmt.Errorf("mock csi driver delete error")}},

huffmanca commented:
@xing-yang ,

I've updated the PR, and the tests pass after following your instructions. I also rebased this PR, as there's been some time since the initial submission.

With that being said, I haven't been able to reproduce the issue in the sidecar, and I'm hesitant about merging it blindly. It seems identical to the fix in the snapshot-controller, which is why it's been submitted.

xing-yang commented:
@huffmanca It is a timing issue, so I think whether it happens depends on the environment.  Your fix looks right to me.

xing-yang commented:
/lgtm
/approve
